src/System.Memory/src/System/Buffers/SequenceReader.cs

```diff
@@ -0,0 +1,373 @@
+        /// </summary>
+        public ReadOnlySpan<T> CurrentSpan { get; private set; }
+
+        /// <summary>
+        /// The index in the <see cref="CurrentSpan"/>.
+        /// </summary>
+        public int CurrentSpanIndex { get; private set; }
+
+        /// <summary>
+        /// The unread portion of the <see cref="CurrentSpan"/>.
+        /// </summary>
+        public ReadOnlySpan<T> UnreadSpan
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => CurrentSpan.Slice(CurrentSpanIndex);
+        }
+
+        /// <summary>
+        /// The total number of <typeparamref name="T"/>'s processed by the reader.
+        /// </summary>
+        public long Consumed { get; private set; }
+
+        /// <summary>
+        /// Remaining <typeparamref name="T"/>'s in the reader's <see cref="Sequence"/>.
+        /// </summary>
+        public long Remaining => Length - Consumed;
+
+        /// <summary>
+        /// Count of <typeparamref name="T"/> in the reader's <see cref="Sequence"/>.
+        /// </summary>
+        public long Length
+        {
+            get
+            {
+                if (_length < 0)
+                {
+                    // Cache the length
+                    _length = Sequence.Length;
+                }
+                return _length;
+            }
+        }
+
+        /// <summary>
+        /// Peeks at the next value without advancing the reader.
+        /// </summary>
+        /// <param name="value">The next value or default if at the end.</param>
+        /// <returns>False if at the end of the reader.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool TryPeek(out T value)
+        {
+            if (_moreData)
+            {
+                value = CurrentSpan[CurrentSpanIndex];
+                return true;
+            }
+            else
+            {
+                value = default;
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Read the next value and advance the reader.
+        /// </summary>
+        /// <param name="value">The next value or default if at the end.</param>
+        /// <returns>False if at the end of the reader.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool TryRead(out T value)
+        {
+            if (End)
+            {
+                value = default;
+                return false;
+            }
+
+            value = CurrentSpan[CurrentSpanIndex];
+            CurrentSpanIndex++;
+            Consumed++;
+
+            if (CurrentSpanIndex >= CurrentSpan.Length)
+            {
+                GetNextSpan();
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Move the reader back the specified number of items.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Rewind(long count)
```

Comment:
It's the right trade-off for Rewind to be forcibly inlined?

Reply:
We believe so. It is only the fast path and is expected to be used frequently in some scenarios. I'll look to push the count check to the slow path like I did for advance to make this a bit better.